internal/database/webhook_logs.go

```diff
@@ -0,0 +1,374 @@
+		encKeyID,
+		sqlf.Join(webhookLogColumns, ", "),
+	)
+
+	row := s.QueryRow(ctx, q)
+	if err := s.scanWebhookLog(ctx, log, row); err != nil {
+		return errors.Wrap(err, "scanning webhook log")
+	}
+
+	return nil
+}
+
+func (s *webhookLogStore) GetByID(ctx context.Context, id int64) (*types.WebhookLog, error) {
+	q := sqlf.Sprintf(
+		webhookLogGetByIDQueryFmtstr,
+		sqlf.Join(webhookLogColumns, ", "),
+		id,
+	)
+
+	row := s.QueryRow(ctx, q)
+	log := types.WebhookLog{}
+	if err := s.scanWebhookLog(ctx, &log, row); err != nil {
+		return nil, errors.Wrap(err, "scanning webhook log")
+	}
+
+	return &log, nil
+}
+
+type WebhookLogListOpts struct {
+	// The maximum number of entries to return, and the cursor, if any. This
+	// doesn't use LimitOffset because we're paging down a potentially changing
+	// result set, so our cursor needs to be based on the ID and not the row
+	// number.
+	Limit  int
+	Cursor int64
+
+	// If set and non-zero, this limits the webhook logs to those matched to
+	// that external service. If set and zero, this limits the webhook logs to
+	// those that did not match an external service. If nil, then all webhook
+	// logs will be returned.
+	ExternalServiceID *int64
+
+	// If set, only webhook logs that resulted in errors will be returned.
+	OnlyErrors bool
+
+	Since *time.Time
+	Until *time.Time
+}
+
+func (opts *WebhookLogListOpts) predicates() []*sqlf.Query {
+	preds := []*sqlf.Query{sqlf.Sprintf("TRUE")}
+	if id := opts.ExternalServiceID; id != nil {
+		if *id == 0 {
+			preds = append(preds, sqlf.Sprintf("external_service_id IS NULL"))
+		} else {
+			preds = append(preds, sqlf.Sprintf("external_service_id = %s", *id))
+		}
+	}
+	if opts.OnlyErrors {
+		preds = append(preds, sqlf.Sprintf("status_code NOT BETWEEN 100 AND 399"))
+	}
+	if since := opts.Since; since != nil {
+		preds = append(preds, sqlf.Sprintf("received_at >= %s", *since))
+	}
+	if until := opts.Until; until != nil {
+		preds = append(preds, sqlf.Sprintf("received_at <= %s", *until))
+	}
+
+	return preds
+}
+
+func (s *webhookLogStore) Count(ctx context.Context, opts WebhookLogListOpts) (int64, error) {
+	q := sqlf.Sprintf(
+		webhookLogCountQueryFmtstr,
+		sqlf.Join(opts.predicates(), " AND "),
+	)
+
+	row := s.QueryRow(ctx, q)
+	var count int64
+	if err := row.Scan(&count); err != nil {
+		return 0, err
+	}
+
+	return count, nil
+}
+
+func (s *webhookLogStore) List(ctx context.Context, opts WebhookLogListOpts) ([]*types.WebhookLog, int64, error) {
+	preds := opts.predicates()
+	if cursor := opts.Cursor; cursor != 0 {
+		preds = append(preds, sqlf.Sprintf("id <= %s", cursor))
+	}
+
+	var limit *sqlf.Query
+	if opts.Limit != 0 {
+		limit = sqlf.Sprintf("LIMIT %s", opts.Limit+1)
+	} else {
+		limit = sqlf.Sprintf("")
+	}
+
+	q := sqlf.Sprintf(
+		webhookLogListQueryFmtstr,
+		sqlf.Join(webhookLogColumns, ", "),
+		sqlf.Join(preds, " AND "),
+		limit,
+	)
+
+	rows, err := s.Query(ctx, q)
+	if err != nil {
+		return nil, 0, err
+	}
+	defer rows.Close()
```

Comment:
missing `rows.Err()` check can be covered by using 
```suggestion
	defer func() {
	err = basestore.CloseRows(rows, err)
	}
```

Reply:
Didn't know that existed; thanks!